File: 380. Insert Delete GetRandom O(1)/Insert Delete GetRandom O(1).js
class RandomizedSet {
    constructor() {
        this.valMap = new Map(),
        this.indexMap = new Map();
        
        this.highest = -1;
    }
    
    insert(val) {
        if(this.valMap.has(val)) return false;
        
        this.highest += 1;
        this.valMap.set(val, this.highest);
        this.indexMap.set(this.highest, val);
        
        return true;
    }
    
    remove(val) {
        if(!this.valMap.has(val)) return false;
        
        const currIdx = this.valMap.get(val);
        const valAtLastIdx = this.indexMap.get(this.highest);
        this.indexMap.set(currIdx, valAtLastIdx); 
        this.valMap.set(valAtLastIdx, currIdx);
        this.valMap.delete(val);
        this.indexMap.delete(this.highest);
        
        this.highest -= 1;
        
        return true;
    }
    
    getRandom() {
        if(this.highest < 0) return null;
        
        const rnd = this.generateRandom();
        return this.indexMap.get(rnd);
    }
    
    generateRandom() {
        return Math.floor(Math.random() * (this.highest + 1));
    }
}

File: 380. Insert Delete GetRandom O(1)/Insert Delete GetRandom O(1).py
class RandomizedSet:
    def __init__(self):
        self.dic_direct = {}
        self.dic_invert = {}
        self.num_elem = 0
        
    def insert(self, val: int) -> bool:
        if val in self.dic_invert:
            return False
        else:
            self.dic_invert[val] = self.num_elem
            self.dic_direct[self.num_elem] = val
            self.num_elem += 1
            return True
        
    def remove(self, val):
        if val not in self.dic_invert:
            return False
        else:
            ind = self.dic_invert.pop(val)
            self.dic_direct.pop(ind)
            if ind != self.num_elem - 1:
                self.dic_direct[ind] = self.dic_direct[self.num_elem - 1]
                self.dic_invert[self.dic_direct[self.num_elem - 1]] = ind
                self.dic_direct.pop(self.num_elem - 1)
            self.num_elem -= 1
            return True
        
    def getRandom(self):
        index = floor(random.random()*self.num_elem)
        return self.dic_direct[index]